contracts/NonfungiblePositionManagerV4.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {IERC721Metadata} from "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {IERC721} from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+
+import {INonfungiblePositionManagerV4} from "./interfaces/INonfungiblePositionManagerV4.sol";
+import {PeripheryValidation} from "./base/PeripheryValidation.sol";
+import {PeripheryPayments} from "./base/PeripheryPayments.sol";
+import {PeripheryImmutableState} from "./base/PeripheryImmutableState.sol";
+import {SelfPermit} from "./base/SelfPermit.sol";
+import {LiquidityManagement} from "./base/LiquidityManagement.sol";
+import {Multicall} from "./base/Multicall.sol";
+
+contract NonfungiblePositionManagerV4 is
+    INonfungiblePositionManagerV4,
+    ERC721,
+    PeripheryImmutableState,
+    PeripheryValidation,
+    LiquidityManagement,
+    SelfPermit,
+    Multicall
+{
+    using PoolIdLibrary for PoolKey;
+
+    error InvalidTokenID();
+    error NotApproved();
+    error NotCleared();
+    error NonexistentToken();
+
+    // details about the Uniswap position
+    struct TokenPosition {
+        // the nonce for permits
+        uint96 nonce;
+        // the address that is approved for spending this token
+        address operator;
```

Comment:
hm this is kind of strange because you can have more than 1 operator with the 721 spec... so not sure why we save just one. did we do the same in v3? if so.. why?